eZ/Publish/Core/Persistence/Legacy/Content/Mapper.php

```diff
@@ -176,64 +187,156 @@ public function convertToStorageValue(Field $field)
+     */
+    private function buildContentObjects(
+        array $contentInfos,
+        array $versionInfos,
+        array $fields,
+        array $missingFieldDefinitions,
+        array $versionedNames
+    ): array {
         $results = [];
+
         foreach ($contentInfos as $contentId => $contentInfo) {
             foreach ($versionInfos[$contentId] as $versionId => $versionInfo) {
                 // Fallback to just main language name if versioned name data is missing
-                if (isset($versionedNameData[$contentId][$versionInfo->versionNo])) {
-                    $names = $versionedNameData[$contentId][$versionInfo->versionNo];
-                } else {
-                    $names = [$contentInfo->mainLanguageCode => $contentInfo->name];
-                }
+                $names = $versionedNames[$contentId][$versionInfo->versionNo]
+                    ?? [$contentInfo->mainLanguageCode => $contentInfo->name];
 
                 $content = new Content();
                 $content->versionInfo = $versionInfo;
                 $content->versionInfo->names = $names;
                 $content->versionInfo->contentInfo = $contentInfo;
                 $content->fields = array_values($fields[$contentId][$versionId]);
+
+                $missingVersionFieldDefinitions = $missingFieldDefinitions[$contentId][$versionId];
+                foreach ($missingVersionFieldDefinitions as $languageCode => $versionFieldDefinitions) {
+                    foreach ($versionFieldDefinitions as $fieldDefinition) {
+                        $content->fields[] = $this->createEmptyField(
+                            $fieldDefinition,
+                            $languageCode
+                        );
+                    }
+                }
+
                 $results[] = $content;
             }
         }
 
         return $results;
     }
 
+    /**
+     * @phpstan-return TVersionedLanguageFieldDefinitionsMap
+     *
+     * @throws \eZ\Publish\API\Repository\Exceptions\NotFoundException
+     */
+    private function loadCachedVersionFieldDefinitionsPerLanguage(
+        array $rows,
+        string $prefix
+    ): array {
+        $fieldDefinitions = [];
+        $contentTypes = [];
+        $allLanguages = $this->loadAllLanguagesWithIdKey();
+
+        foreach ($rows as $row) {
+            $contentId = (int)$row["{$prefix}id"];
+            $versionId = (int)$row["{$prefix}version_id"];
+            $contentTypeId = (int)$row["{$prefix}contentclass_id"];
+            $languageMask = (int)$row["{$prefix}version_language_mask"];
+
+            if (isset($fieldDefinitions[$contentId][$versionId])) {
+                continue;
+            }
+
+            $languageCodes = $this->extractLanguageCodesFromMask($languageMask, $allLanguages);
+            $contentType = $contentTypes[$contentTypeId] = $contentTypes[$contentTypeId]
+                ?? $this->contentTypeHandler->load($contentTypeId);
```

Comment:
I was looking at what we are doing here. I had to look into operator precedence, since I didn't immediately recall if assignment operator has lower priority over null-coalescing.

https://www.php.net/manual/en/language.operators.precedence.php

```suggestion
            $contentTypes[$contentTypeId] ??= $this->contentTypeHandler->load($contentTypeId);
            $contentType = $contentTypes[$contentTypeId];
```
or
```suggestion
            $contentType = $contentTypes[$contentTypeId] ??= $this->contentTypeHandler->load($contentTypeId);
```